cmd/metrics.go

```diff
@@ -39,91 +41,130 @@ var (
-	// Increment server uptime
-	serverUptime.Set(UTCNow().Sub(globalBootTime).Seconds())
+// newMinioCollector describes the collector
+// and returns reference of minioCollector
+// It creates the Prometheus Description which is used
+// to define metric and  help string
+func newMinioCollector() *minioCollector {
+	return &minioCollector{
+		desc: prometheus.NewDesc("minio_stats", "Statistics exposed by Minio server", nil, nil),
+	}
+}
+
+// minioCollector is the Custom Collector
+type minioCollector struct {
+	desc *prometheus.Desc
+}
+
+// Describe sends the super-set of all possible descriptors of metrics
+func (c *minioCollector) Describe(ch chan<- *prometheus.Desc) {
+	ch <- c.desc
+}
+
+// Collect is called by the Prometheus registry when collecting metrics.
+func (c *minioCollector) Collect(ch chan<- prometheus.Metric) {
 
 	// Fetch disk space info
 	objLayer := newObjectLayerFn()
 	// Service not initialized yet
 	if objLayer == nil {
 		return
 	}
-
-	// Update total/free disk space
 	s := objLayer.StorageInfo(context.Background())
-	minioStorageTotal.Set(float64(s.Total))
-	minioStorageFree.Set(float64(s.Free))
 
-	// Update online/offline disks
-	onlineMinioDisks.Set(float64(s.Backend.OnlineDisks))
-	offlineMinioDisks.Set(float64(s.Backend.OfflineDisks))
+	// Network Sent/Received Bytes
+	ch <- prometheus.MustNewConstMetric(
+		prometheus.NewDesc(
+			prometheus.BuildFQName("minio", "", "network_sent_bytes_total"),
+			"Total number of bytes sent by current Minio server instance",
+			nil, nil),
+		prometheus.CounterValue,
+		float64(globalConnStats.getTotalOutputBytes()),
+	)
+	ch <- prometheus.MustNewConstMetric(
+		prometheus.NewDesc(
+			prometheus.BuildFQName("minio", "", "network_received_bytes_total"),
+			"Total number of bytes received by current Minio server instance",
+			nil, nil),
+		prometheus.CounterValue,
+		float64(globalConnStats.getTotalInputBytes()),
+	)
+	// Total/Free Storage Bytes
+	ch <- prometheus.MustNewConstMetric(
+		prometheus.NewDesc(
+			prometheus.BuildFQName("minio", "disk", "storage_bytes"),
+			"Total disk storage available to current Minio server instance",
+			nil, nil),
+		prometheus.GaugeValue,
+		float64(s.Total),
+	)
+	ch <- prometheus.MustNewConstMetric(
+		prometheus.NewDesc(
+			prometheus.BuildFQName("minio", "disk", "storage_free_bytes"),
+			"Total free disk storage available to current Minio server instance",
+			nil, nil),
+		prometheus.GaugeValue,
+		float64(s.Free),
+	)
+	// Set Server Start Time
+	ch <- prometheus.MustNewConstMetric(
+		prometheus.NewDesc(
+			prometheus.BuildFQName("minio", "server", "start_time_seconds"),
+			"Time when current Minio server instance started",
+			nil, nil),
+		prometheus.GaugeValue,
+		float64(globalBootTime.Unix()),
+	)
 
-	// Update prometheus metric
-	networkSentBytes.Set(float64(globalConnStats.getTotalOutputBytes()))
-	networkReceivedBytes.Set(float64(globalConnStats.getTotalInputBytes()))
+	// Minio Total Disk/Offline Disk
```

Comment:
Although offline/online disks don't make sense in Minio FS mode, I recommend to show a value other than zero for total disks. The following diff fixes the problem:

```diff
diff --git a/cmd/metrics.go b/cmd/metrics.go
index 942c247d..66c3cea6 100644
--- a/cmd/metrics.go
+++ b/cmd/metrics.go
@@ -124,6 +124,15 @@ func (c *minioCollector) Collect(ch chan<- prometheus.Metric) {
 		float64(globalBootTime.Unix()),
 	)
 
+	var totalDisks, offlineDisks int
+	if s.Backend.Type == FS {
+		totalDisks = 1
+		offlineDisks = 0
+	} else {
+		offlineDisks = s.Backend.OfflineDisks
+		totalDisks = s.Backend.OfflineDisks + s.Backend.OnlineDisks
+	}
+
 	// Minio Total Disk/Offline Disk
 	ch <- prometheus.MustNewConstMetric(
 		prometheus.NewDesc(
@@ -131,7 +140,7 @@ func (c *minioCollector) Collect(ch chan<- prometheus.Metric) {
 			"Total number of disks for current Minio server instance",
 			nil, nil),
 		prometheus.GaugeValue,
-		float64(s.Backend.OnlineDisks+s.Backend.OfflineDisks),
+		float64(totalDisks),
 	)
 	ch <- prometheus.MustNewConstMetric(
 		prometheus.NewDesc(
@@ -139,7 +148,7 @@ func (c *minioCollector) Collect(ch chan<- prometheus.Metric) {
 			"Total number of offline disks for current Minio server instance",
 			nil, nil),
 		prometheus.GaugeValue,
-		float64(s.Backend.OfflineDisks),
+		float64(offlineDisks),
 	)
 }
 ```

Reply:
Done